netcdf4:init at 1.5.1.2

mamueller opened the pull request:
#67337
Generalized  the expression to be able to create a version with parallel
IO. If the function is called with netcdf-mpi, hdf5-mpi and mpi4py
it now produces a version that can access netcdf4 files in parallel.
I tested this with a small python script that uses parallel IO.
This should probably go to the original testsuite of the package which
is disabled in the nix expression I started with.
So I did not touch this part.
The patch is necessary since the original setup.py does not find
the parallel version of the netcdf although it is there since it
inspects the header file with a pattern that is propably only present in
the newest version. I made a pullrequest to the upstream maintainers to
generelize the pattern so that the patch will no longer be needed.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

mamueller committed:
netcdf4:init at 1.5.1.2

Generalized  the expression to be able to create a version with parallel
IO. If the function is called with netcdf-mpi, hdf5-mpi and mpi4py
it now produces a version that can access netcdf4 files in parallel.
I tested this with a small python script that uses parallel IO.
This should probably go to the original testsuite of the package which
is disabled in the nix expression I started with.
So I did not touch this part.
The patch is necessary since the original setup.py does not find
the parallel version of the netcdf although it is there since it
inspects the header file with a pattern that is propably only present in
the newest version. I made a pullrequest to the upstream maintainers to
generelize the pattern so that the patch will no longer be needed.

mamueller committed:
netcdffortran-mpi: new override

This is just a suggested override of netcdffortran wiht hdf5-mpi and netcdf-mpi as inputs.
The resulting fortran library can then be used for parallel IO.
Since I am new to nix it helped me a lot to find the netcdf-mpi package
directly before I was aware of the fact that it actually uses the same
function as netcdf just with the parallel version of hdf5. I therefore
suggest the same for the fortran binding.

lsix commented:
Looks good to me, I'll give it a try when possible (it might take time).

Did you try to see if the tests can be re-enabled?

lsix commented:
The commits should be renamed:

`netcdf4:init at 1.5.1.2` -> `pythonPackages.netcdf4: enable parallel IO support` or something like that.

I would also say that your second commit (introducing `netcdffortran-mpi`) could deserve its own PR, since it is independent (about another package, not dependent on the python netcdf binding).

lsix commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,73 @@
+      inherit mpi;
+    }; 
+    meta = with stdenv.lib; {
+      description = "Interface to netCDF library (versions 3 and 4) with support for parallel IO";
```
I do not think the description should change from the previous implementation.

lsix commented:
And by the way, thanks for your first PR to `nixpkgs` and welcome in the community !

FRidh commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,73 @@
-  pname = "netCDF4";
-  version = "1.5.1.2";
+, numpy, zlib, netcdf, hdf5, curl, libjpeg, cython, cftime 
+,mpi4py ? null, openssh ? null}:
```
space

mamueller replied:
I overlooked this request. So I removed the extra indentation and rebased again.
I have to admit that I am not too familiar with the pull requests workflow, so I apologize for any confusion I might have created by going about it in a roundabout way.

mamueller replied:
Is there more to be done?

FRidh commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,73 @@
-  disabled = isPyPy;
+let 
+  mpiSupport = (mpi4py!=null);
+  mpi= if mpiSupport then mpi4py.mpi else null;
```
space

FRidh commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,73 @@
-  };
+    patches=[./parallel4Detection.patch];
+  
+    preBuild= if mpiSupport then ''
```
`makeFlagsArray` can be set directly. Use `lib.optionals` to extend when `mpiSupport`.

mamueller replied:
I would like to address this but I am not sure how (I do not know what is in the makeFlagsArray originally I was just happy to add the required lines)
I would like to know what I should have done to learn.

veprbl replied:
```suggestion
    makeFlags = lib.optionals mpiSupport [ "CC=mpicc" "CXX=mpicxx" ];
```

FRidh reviewed (changes requested):
Please undo the indentation change that is not needed.

veprbl commented:
@mamueller Are you interested in finishing this?

mamueller commented:
What would have to be done?


Am Fr., 27. Dez. 2019 um 22:27 Uhr schrieb Dmitry Kalinkin <
notifications@github.com>:

> @mamueller <https://github.com/mamueller> Are you interested in finishing
> this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/67338?email_source=notifications&email_token=AAIRGPPAGBP2X6IYY27ADRTQ2ZXLXA5CNFSM4IO7CKDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHXZFCQ#issuecomment-569348746>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIRGPN4UL5KECRHVS55R7DQ2ZXLXANCNFSM4IO7CKDA>
> .
>

veprbl commented:
@mamueller You need to rebase your changes on top of our master branch to fix the merge conflict. Since we already updated to version 1.5.3, you should not need your parallel4Detection.patch anymore [1]. Also, you would need to address the review comments above.

https://github.com/Unidata/netcdf4-python/commit/43ba3e54011d8e0dd3e055b069597eb92ffcf4bc

mamueller commented:
I am a little under pressure just now. So I would not mind if I would not
have to do it right now.
(At the moment I am not working on any nix related stuff, so its a bit
awkward to sell to my boss, while when I made the change we really needed
it ...)

Am Di., 31. Dez. 2019 um 18:55 Uhr schrieb Dmitry Kalinkin <
notifications@github.com>:

> @mamueller <https://github.com/mamueller> You need to rebase your changes
> on top of our master branch to fix the merge conflict. Since we already
> updated to version 1.5.3, you should not need your parallel4Detection.patch
> anymore [1]. Also, you would need to address the review comments above.
>
> Unidata/netcdf4-python@43ba3e5
> <https://github.com/Unidata/netcdf4-python/commit/43ba3e54011d8e0dd3e055b069597eb92ffcf4bc>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/67338?email_source=notifications&email_token=AAIRGPN7JJIN7BJZWIMHJDTQ3OBQ3A5CNFSM4IO7CKDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH4P6JA#issuecomment-569966372>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIRGPMD5LUQ4MNHPDWM3GLQ3OBQ3ANCNFSM4IO7CKDA>
> .
>

mamueller committed:
updated my previous extension of the netcdf4 expression #67338.
Now also the tests run except one that is skipped by the patch.
The behavior of the tests is puzzeling to me in two ways:
1. If I drop into nix-shell (even  with --pure ) and call genericBuild
   all tests run but if I use nix-build the test fails.
   (If you look at the test you see that it tries to load a remote url.
   I have no clue why this works in the shell and not in nix-build)

2. A similar thing happens with py.test:
   It works in the nix-shell but fails with nix-build but here the test
   suite does not even start.
   I used unittest instead which leads to only one test failing (which
   is now skipped by the patch)

mamueller commented:
Sorry for the late response.
I just had a little time to work on it and updated to the new netcdf4 version.
Even (most) tests run now.
With some strange things that I cannot figure out as described in my commit message.
I am not sure how to address the request by FRidh, because I am still (very) new to nix and not too familiar with the stdenv. I think it will be a one liner.
Hopefully the pullrequest goes through now.
Cheers and thanks for the friendly welcome.
Markus

jonringer commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,72 @@
+      sha256 = "2a3ca855848f4bbf07fac366da77a681fcead18c0a8813d91d46302f562dc3be";
+    };
+  
+    checkInputs = [ netcdf numpy ] ++(if mpiSupport then [ mpi4py openssh mpi] else []);
```
```suggestion
    checkInputs = [ netcdf numpy ] 
      ++ lib.optionals mpiSupport [ mpi4py openssh mpi ];
```

veprbl replied:
I doubt that `openssh` and `mpi` need to be in the `checkInputs` in the first place.

mamueller replied:
If I don't add  openssh I get the following error while running the tests:
-------------------------------------------------------------------------
The value of the MCA parameter "plm_rsh_agent" was set to a path
that could not be found:

  plm_rsh_agent: ssh : rsh

Please either unset the parameter, or check that the path is correct
--------------------------------------------------------------------------
[localhost:00218] [[INVALID],INVALID] FORCE-TERMINATE AT Not found:-13 - error plm_rsh_component.c(335)
[localhost:00217] [[INVALID],INVALID] ORTE_ERROR_LOG: Unable to start a daemon on the local node in file ess_singleton_module.c at line 716
[localhost:00217] [[INVALID],INVALID] ORTE_ERROR_LOG: Unable to start a daemon on the local node in file ess_singleton_module.c at line 172
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  orte_ess_init failed
  --> Returned value Unable to start a daemon on the local node (-127) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_mpi_init: ompi_rte_init failed
  --> Returned "Unable to start a daemon on the local node" (-127) instead of "Success" (0)
--------------------------------------------------------------------------
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)

but I removed mpi which was indeed a leftover

jonringer commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,72 @@
+  
+    checkInputs = [ netcdf numpy ] ++(if mpiSupport then [ mpi4py openssh mpi] else []);
+  
+    buildInputs = [
+      cython 
+      mpi 
+    ];
```
cython needs to be executed, should go in nativebuildinputs
```suggestion
    nativeBuildInputs = [
      cython
    ];
    
    buildInputs = [
      mpi 
    ];
```

jonringer commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,72 @@
+      hdf5
+      curl
+      libjpeg
+    ] ++ (if mpiSupport then [mpi4py mpi openssh] else []);
```
```suggestion
    ] ++ lib.optionals mpiSupport [ mpi4py mpi openssh ];
```

jonringer commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,72 @@
-    inherit pname version;
-    sha256 = "2a3ca855848f4bbf07fac366da77a681fcead18c0a8813d91d46302f562dc3be";
-  };
+    patches=[./skipDapTest.patch];
```
spacing
```suggestion
    patches = [ ./skipDapTest.patch ];
```

jonringer commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,72 @@
+    passthru = {
+      mpiSupport = mpiSupport;
+      inherit mpi;
+    }; 
```
```suggestion
    }; 

```

jonringer commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,72 @@
+    }; 
+    meta = with stdenv.lib; {
+      description = "Interface to netCDF library (versions 3 and 4)";
+      homepage = https://pypi.python.org/pypi/netCDF4;
+      license = licenses.free;  # Mix of license (all MIT* like)
```
please add yourself as a maintainer

jonringer commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,72 @@
+    }; 
+    meta = with stdenv.lib; {
+      description = "Interface to netCDF library (versions 3 and 4)";
+      homepage = https://pypi.python.org/pypi/netCDF4;
```
rfc 45 happened
```suggestion
      homepage = "https://pypi.python.org/pypi/netCDF4";
```

veprbl commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,50 +1,72 @@
+    NETCDF4_DIR="${netcdf}";
+    CURL_DIR="${curl.dev}";
+    JPEG_DIR="${libjpeg.dev}";
+    #  test for mpi first and then set
+    # MPI_INCDIR="${mpi}/include" ;
+    # is not necessary
```
Why do you add these three lines?

mamueller replied:
Accident...

veprbl commented on pkgs/development/python-modules/netcdf4/skipDapTest.patch:
```diff
@@ -0,0 +1,60 @@
+Only in netCDF4-1.5.3.new: nix_run_setup
+Only in netCDF4-1.5.3.new/test: __pycache__
+diff -cr netCDF4-1.5.3/test/tst_dap.py netCDF4-1.5.3.new/test/tst_dap.py
+*** netCDF4-1.5.3/test/tst_dap.py	2019-10-21 18:38:12.000000000 +0200
```
This patch is messy. Please remove it. You probably should be able to achieve the same effect with "rm test/tst_dap.py". Would be better to switch to `pytestCheckHook` and put those in `disabledTests`.

mamueller replied:
I have used the rm version, since I had trouble using pytest (which I tried first).
I observed some very strange behavior that I could not understand at all:
When I used nix-shell --pure the pytest was working but with nix-build they never got further than starting to collect. I could not see the actual error message. 
This is the reason I switched to unittest.
This is very esoteric and for my further endeavors in the nix world I would be very happy to get a clue what exactly is different in the nix-shell env. I have hunted for shell variables but could not explain the different outcome.

mamueller replied:
Sorry to bother you.
I am not sure how to proceed. I think I have incorporated everything that had been demanded.
I would like to finish this.
Is there something still blocking?
Thanks in advance 
Markus

mamueller committed:
Update pkgs/development/python-modules/netcdf4/default.nix

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

mamueller committed:
hopefully addressed all the comments

mamueller committed:
removed the added indentation

mamueller committed:
added the desired quotation marks

veprbl commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -26,21 +39,30 @@ buildPythonPackage rec {
+  
+  passthru = {
+    mpiSupport = mpiSupport;
+    inherit mpi;
```
```suggestion
    inherit mpi mpiSupport;
```
and remove previous line

veprbl reviewed (changes requested):
@mamueller Yes there are a few things you missed from previous review. This is looking better (style-wise), but there are still few rough edges to fix. Please see my comments below.

mamueller committed:
Update pkgs/top-level/all-packages.nix

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

mamueller committed:
Update pkgs/development/python-modules/netcdf4/default.nix

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

mamueller committed:
Update pkgs/development/python-modules/netcdf4/default.nix

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

mamueller committed:
Update pkgs/development/python-modules/netcdf4/default.nix

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

mamueller committed:
Update pkgs/development/python-modules/netcdf4/default.nix

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

mamueller committed:
Update pkgs/development/python-modules/netcdf4/default.nix

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

mamueller committed:
Incorporated all the remaining changes

mamueller commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -26,21 +39,30 @@ buildPythonPackage rec {
+  HDF5_DIR= lib.getDev hdf5;
+  NETCDF4_DIR="${netcdf}";
+  CURL_DIR="${curl.dev}";
+  JPEG_DIR="${libjpeg.dev}";
```
Thanks for your time. I learned a bit about nix (I can now use inherit properly ;-) and also see the point of lib.optionals..)
I have committed all the suggestions and added the remaining outdated ones manually.
I have some remaining questions:

1. I am sorry for having created such a lot of style problems.
   Is there a linter (or at least a style guide)?
   I am using vim with vim-nix 
   For python code I use the python language server integration which is very verbose
   Is there something similar available for nix?
2. I am a bit unsure about the workflow used to deal with pull requests. 
( I normally work in a very small group at a research institute where 99.999 % of git usage consists of pull and push and even branches are an exception..., so we are not very professional users..)
 I know what rebasing means and did it twice already for this branch, but  I do not know for instance if you would prefer my (many tiny) commits to be squashed into one or if I would make things harder by it...)
So any suggestions about using git professionally are also welcome.

Thanks again.

veprbl replied:
@mamueller 
Regarding your questions:

1. There are couple linters that could help. `nixpkgs-fmt` is one of the most promising ones. There are several other ones: https://github.com/nix-community/nixpkgs-fmt#formatters
Note that if will format whole file at the time, which may hinder the review process if you mix functional changes with style changes. A typical solution to that is to format the original file in a first commit and then introduce functional changes in subsequent separate commits. It is a bit too late to do that in this PR, but you could try this in the future.

2. Usually commits are supposed to have one specific idea for a change that is well described in the commit message. So if a PR contains several separate changes, we usually ask for them to be in separate commits. In case of this PR there is only one change  (adding mpi support to pythonPackages.netcdf4), so we might just squash everything on merge for you. If you are unsure what to do, look around at other peoples contributions, see what they do.

FRidh commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -26,21 +40,29 @@ buildPythonPackage rec {
-    py.test test/tst_*.py
+    cd test
+    rm tst_dap.py # does not fail in the nix-shell  
+    python -m unittest tst_*.py
```
`${python.interpreter}` instead of `python`. This is a prerequisite for supporting non-CPython interpreters such as PyPy.

veprbl commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,21 +1,35 @@
 
   disabled = isPyPy;
-
+  
```
This change is not needed

veprbl commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -1,21 +1,35 @@
 
   buildInputs = [
-    cython
+    mpi 
```
```suggestion
    mpi
```

veprbl commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -26,21 +40,29 @@ buildPythonPackage rec {
-  ];
+  ] ++ lib.optionals mpiSupport [mpi4py mpi openssh];
+
+  #patches=[./skipDapTest.patch];
```
Please remove this

veprbl commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -26,21 +40,29 @@ buildPythonPackage rec {
-  NETCDF4_DIR=netcdf;
-  CURL_DIR=curl.dev;
-  JPEG_DIR=libjpeg.dev;
+  NETCDF4_DIR="${netcdf}";
```
```suggestion
  NETCDF4_DIR=netcdf;
```

veprbl commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -26,21 +40,29 @@ buildPythonPackage rec {
-  CURL_DIR=curl.dev;
-  JPEG_DIR=libjpeg.dev;
+  NETCDF4_DIR="${netcdf}";
+  CURL_DIR="${curl.dev}";
```
```suggestion
  CURL_DIR=curl.dev;
```

veprbl commented on pkgs/development/python-modules/netcdf4/default.nix:
```diff
@@ -26,21 +40,29 @@ buildPythonPackage rec {
-  JPEG_DIR=libjpeg.dev;
+  NETCDF4_DIR="${netcdf}";
+  CURL_DIR="${curl.dev}";
+  JPEG_DIR="${libjpeg.dev}";
```
```suggestion
  JPEG_DIR=libjpeg.dev;
```

veprbl reviewed (commented):
Noticed there were couple whitespace issues. Also, reposting some previous comments that were not addressed.

FRidh commented:
closes https://github.com/NixOS/nixpkgs/issues/67340.

jonringer reviewed (approved):
assuming other reviews are addressed
```
https://github.com/NixOS/nixpkgs/pull/67338
1 package built:
netcdffortran-mpi
```

cc @bhipple  for mpi conventions

bhipple commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5381,6 +5381,11 @@ in
+  netcdffortran-mpi = appendToName "mpi" (netcdffortran.override {
+    hdf5 = hdf5-mpi;
+    netcdf = netcdf-mpi;
+  });
```
We haven't done anything with `mpi`, but the direction we went with blas/lapack was to have 1 variable at the top level specifying your variant, and then an overlay can -- in one place -- consistently override everything to be the right flavor.

This ensures you don't end up with two incoherent flavors of something like numpy in your stack.  CC @matthewbauer not sure if we should do something similar for `mpi`.

jonringer replied:
ah, i did get them mixed up. But I think we should have similar conventions between blas and mpi.
